exonum-java-binding/core/src/main/java/com/exonum/binding/core/runtime/ServiceRuntime.java

```diff
@@ -51,99 +63,279 @@
 
   private static final Logger logger = LogManager.getLogger(ServiceRuntime.class);
 
-  private final Injector frameworkInjector;
   private final ServiceLoader serviceLoader;
+  private final ServicesFactory servicesFactory;
+  /**
+   * The list of services. It is stored in a sorted map that offers the same iteration order
+   * on all nodes with the same services, which is required for correct beforeCommit
+   * operation.
+   */
+  private final SortedMap<String, ServiceWrapper> services;
+  private final Map<Integer, ServiceWrapper> servicesById;
   private final Object lock = new Object();
 
   /**
    * Creates a new runtime with the given framework injector. Starts the server on instantiation;
    * never stops it.
    *
-   * @param frameworkInjector the injector that has been configured with the Exonum framework
-   *     bindings. It serves as a parent for service injectors
    * @param serviceLoader a loader of service artifacts
+   * @param servicesFactory the factory of services
    * @param server a web server providing transport to Java services
    * @param serverPort a port for the web server providing transport to Java services
    */
   @Inject
-  public ServiceRuntime(Injector frameworkInjector, ServiceLoader serviceLoader, Server server,
+  public ServiceRuntime(ServiceLoader serviceLoader, ServicesFactory servicesFactory, Server server,
       @Named(SERVICE_WEB_SERVER_PORT) int serverPort) {
-    this.frameworkInjector = checkNotNull(frameworkInjector);
     this.serviceLoader = checkNotNull(serviceLoader);
+    this.servicesFactory = checkNotNull(servicesFactory);
+    services = new TreeMap<>();
```

Comment:
I would move instantiations to the field ^

Reply:
What is the reasoning behind your suggestion? I can think of bringing the definition closer to the field declaration 🤔

Reply:
exactly